### PR TITLE
interfaces/builtin: add intel realsense udev rules into camera interface

### PR DESCRIPTION
**Pull request by swem**
As discussed in #2236 , add intel relasense udev rules into camera interface instead of creating a custom interface.
https://github.com/snapcore/snapd/pull/2236#issuecomment-278917634

### CODE REVIEW

File: interfaces/builtin/camera.go [hunk 1]
```diff
@@ -34,11 +36,65 @@ const cameraConnectedPlugAppArmor = `
...
-		reservedForOS:         true,
+var cameraPermanentSlotUdev = []byte(`
+# RealSense UVC cameras (R200, F200, SR300 LR200, ZR300)
+SUBSYSTEMS=="usb", ATTRS{idVendor}=="8086", ATTRS{idProduct}=="0a80", MODE:="0666", GROUP:="plugdev"
+SUBSYSTEMS=="usb", ATTRS{idVendor}=="8086", ATTRS{idProduct}=="0a66", MODE:="0666", GROUP:="plugdev"
+SUBSYSTEMS=="usb", ATTRS{idVendor}=="8086", ATTRS{idProduct}=="0aa5", MODE:="0666", GROUP:="plugdev"
+SUBSYSTEMS=="usb", ATTRS{idVendor}=="8086", ATTRS{idProduct}=="0abf", MODE:="0666", GROUP:="plugdev"
+SUBSYSTEMS=="usb", ATTRS{idVendor}=="8086", ATTRS{idProduct}=="0acb", MODE:="0666", GROUP:="plugdev"
+SUBSYSTEMS=="usb", ATTRS{idVendor}=="8086", ATTRS{idProduct}=="0ad0", MODE:="0666", GROUP:="plugdev"
+SUBSYSTEMS=="usb", ATTRS{idVendor}=="8086", ATTRS{idProduct}=="04b4", MODE:="0666", GROUP:="plugdev"
```
Comment (jdstrand) [inline interfaces/builtin/camera.go]: On my classic system I see:
```
$ ls -l /dev/video0 
crw-rw----+ 1 root video 81, 0 Feb 23 06:53 /dev/video0

$ udevadm info /dev/video0
P: /devices/pci0000:00/0000:00:14.0/usb1/1-5/1-5:1.0/video4linux/video0
N: video0
S: v4l/by-id/usb-CN09GTFM7248767IB6A8A01_Integrated_Webcam_HD-video-index0
S: v4l/by-path/pci-0000:00:14.0-usb-0:5:1.0-video-index0
E: COLORD_DEVICE=1
E: COLORD_KIND=camera
E: DEVLINKS=/dev/v4l/by-path/pci-0000:00:14.0-usb-0:5:1.0-video-index0 /dev/v4l/by-id/usb-CN09GTFM7248767IB6A8A01_Integrated_Webcam_HD-video-index0
E: DEVNAME=/dev/video0
E: DEVPATH=/devices/pci0000:00/0000:00:14.0/usb1/1-5/1-5:1.0/video4linux/video0
E: ID_BUS=usb
E: ID_FOR_SEAT=video4linux-pci-0000_00_14_0-usb-0_5_1_0
E: ID_MODEL=Integrated_Webcam_HD
E: ID_MODEL_ENC=Integrated_Webcam_HD
E: ID_MODEL_ID=670c
E: ID_PATH=pci-0000:00:14.0-usb-0:5:1.0
E: ID_PATH_TAG=pci-0000_00_14_0-usb-0_5_1_0
E: ID_REVISION=5626
E: ID_SERIAL=CN09GTFM7248767IB6A8A01_Integrated_Webcam_HD
E: ID_TYPE=video
E: ID_USB_DRIVER=uvcvideo
E: ID_USB_INTERFACES=:0e0100:0e0200:
E: ID_USB_INTERFACE_NUM=00
E: ID_V4L_CAPABILITIES=:capture:
E: ID_V4L_PRODUCT=Integrated_Webcam_HD
E: ID_V4L_VERSION=2
E: ID_VENDOR=CN09GTFM7248767IB6A8A01
E: ID_VENDOR_ENC=CN09GTFM7248767IB6A8A01
E: ID_VENDOR_ID=0c45
E: MAJOR=81
E: MINOR=0
E: SUBSYSTEM=video4linux
E: TAGS=:uaccess:seat:
E: USEC_INITIALIZED=13275962
```

and if I plugin a usb webcam I get:
```
$ ls -l /dev/video1
crw-rw----+ 1 root video 81, 1 Feb 23 07:11 /dev/video1

$ udevadm info /dev/video1
P: /devices/pci0000:00/0000:00:14.0/usb1/1-2/1-2:1.0/video4linux/video1
N: video1
S: v4l/by-id/usb-046d_09c1-video-index0
S: v4l/by-path/pci-0000:00:14.0-usb-0:2:1.0-video-index0
E: COLORD_DEVICE=1
E: COLORD_KIND=camera
E: DEVLINKS=/dev/v4l/by-id/usb-046d_09c1-video-index0 /dev/v4l/by-path/pci-0000:00:14.0-usb-0:2:1.0-video-index0
E: DEVNAME=/dev/video1
E: DEVPATH=/devices/pci0000:00/0000:00:14.0/usb1/1-2/1-2:1.0/video4linux/video1
E: ID_BUS=usb
E: ID_FOR_SEAT=video4linux-pci-0000_00_14_0-usb-0_2_1_0
E: ID_MODEL=09c1
E: ID_MODEL_ENC=09c1
E: ID_MODEL_ID=09c1
E: ID_PATH=pci-0000:00:14.0-usb-0:2:1.0
E: ID_PATH_TAG=pci-0000_00_14_0-usb-0_2_1_0
E: ID_REVISION=0005
E: ID_SERIAL=046d_09c1
E: ID_TYPE=video
E: ID_USB_DRIVER=uvcvideo
E: ID_USB_INTERFACES=:0e0100:0e0200:010100:010200:
E: ID_USB_INTERFACE_NUM=00
E: ID_V4L_CAPABILITIES=:capture:
E: ID_V4L_PRODUCT=UVC Camera (046d:09c1)
E: ID_V4L_VERSION=2
E: ID_VENDOR=046d
E: ID_VENDOR_ENC=046d
E: ID_VENDOR_ID=046d
E: MAJOR=81
E: MINOR=1
E: SUBSYSTEM=video4linux
E: TAGS=:seat:uaccess:
E: USEC_INITIALIZED=1080970387
```

Shouldn't the mode be 0660 and group video? Why did you choose these? What happens if you don't set either? I see in https://github.com/IntelRealSense/librealsense/blob/master/config/99-realsense-libusb.rules that they use the entries you have, but this is not inline with how v4l is typically setup. On an all snaps system without console users, this is less of a concern, but on a classic system, it is. Can you comment?

You might be interested in this bug for reference: https://bugs.launchpad.net/snappy/+bug/1646144

Comment (jdstrand) [inline interfaces/builtin/camera.go]: Not that the acl package is installed in core (see https://bugs.launchpad.net/snappy/+bug/1646144), I think you can use 0660 and video. If realsense is v4l, then 50-udev-default.rules should add it automatically to the 'video' group and 70-uaccess.rules should add it to uaccess, which means you shouldn't need GROUP at all. Please check if you need MODE (if you do, set it to 0660). With this and the uaccess rule, this should work consistently with how other usb cameras are setup: root and console users can access it. non-root, non-console users would need something along the lines of what I put forth in https://bugs.launchpad.net/snappy/+bug/1646144, but right now that is an unsupported corner case in snappy (services in snappy always run as root), so fixing that shouldn't block this PR.

Comment (swem) [inline interfaces/builtin/camera.go]: @jdstrand I agree the group for realsense camera might be better to align with v4l conventional setup. About the acl mode, librealsense use libusb to claim realsense motion module usb interface https://github.com/IntelRealSense/librealsense/blob/v1.12.1/src/uvc-v4l2.cpp#L524  I've verified that acl mode 666 is still needed.
I'll update the pull request accordingly. Thanks your help!.

**Comment by niemeyer** [comment]
@swem Is this ready for another look from @jdstrand?

**Comment by swem** [comment]
@niemeyer Yes. It's ready for another look from @jdstrand 
Simply remove plugdev group from udev rule.

**Comment by jdstrand** [comment]
Thank you for making the plugdev changes, but I'm still uncomfortable with making the devices 0666. This interface is an implicit core interface which means that there PermanentSlot udev rules will be on every device where snapd is installed and if the usb device is plugged in the cameras can be read by any unconfined process, which is a problem on classic. These realsense cameras should IMO work exactly like any other v4l camera with the same permissions. Right now, those cameras can be accessed via sudo or daemons, which is the correct behavior considering the current limitations of ACLs in snapd.

If you change this to 0660 then +1. Otherwise -1. I suggest committing as 0660 and then working with the snappy team to get what I outlined in https://github.com/snapcore/snapd/pull/2855#discussion_r102711567 prioritized and implemented. I'd be happy to participate in design review of that feature.

**Comment by swem** [comment]
@jdstrand I agree that those camera and motion sensor can be accessed via sudo or daemons. But this does not apply to GUI applications.

Right now I've already upload the librealsense snap onto the store. After connecting camera and raw-usb interfaces manually, command line based utilities works with sudo. Such as:
sudo librealsense.cpp-enumerate-devices

However, GUI application doesn't works with sudo. Such as
sudo librealsense.cpp-config-ui

The error message: No protocol specified.
I think it's related to X environment not being forwarded into snap confinement.

Is there any suggestion on executing GUI snap application via sudo ?

**Comment by niemeyer** [comment]
@swem I believe @jdstrand already suggested an alternative in the comment he linked above.

**Comment by swem** [comment]
Currently udev already create Intel Realsense device node (both v4l node and usb bus node) with 0660 acl permission. I think we don't need to explicit add 066 acl permission.

Now that the Intel's udev rule couldn't be merge into camera interface. I'll keep track on the https://bugs.launchpad.net/snappy/+bug/1646144 to see if Intel realsense camera could leverage it and allow unprivileged user to execute librealsense applications.

**Comment by jdstrand** [comment]
'sudo gui-app' should work ok-- I suspect you need to specify '-H' to sudo to that HOME is set correctly. Does this work:
```
sudo -H librealsense.cpp-config-ui
```

?

**Comment by swem** [comment]
@jdstrand It doesn't work. Same error could be found on executing ubuntu-clock-app snap with sudo.

opengl and unity7 interfaces have already been connected (by default).

**Pull request closed without merging**